[DONOTMERGE] Backport Dynamic Allocation on Kube with Graceful Decommissioning

holdenk opened the pull request:
Do not merge. This is for information purposes for any other organizations that may be needing to back port Dynamic Allocation on Kube & Graceful Decommissioning to their internal Spark 2.4.X

holdenk committed:
Start backporting DA/DS to 2.4. Pre-surgery changes

holdenk committed:
More porting work

holdenk committed:
Map ids are ints in 2.4

holdenk committed:
Fix up some imports

holdenk committed:
First part of backport compiles yay

holdenk committed:
Get tests compiling

holdenk committed:
We broadened isShuffle as part of the backport

holdenk committed:
Bring back some tests for good luck

holdenk committed:
Add the unit suite

holdenk committed:
Work on debugging the failing integration tests

holdenk committed:
Allow runnign int tests without package

holdenk committed:
Cleanup backported decom int test.

holdenk committed:
Fix up setting up the integration test env to not need package

holdenk committed:
Add decommissioning python script

holdenk committed:
More time plz

holdenk committed:
Register signal handler, bump up log level while waiting to shutdown.

holdenk committed:
Add workerDecommissioning to Kube conf

holdenk committed:
Add decom script to life cycle in Kube if enabled

holdenk committed:
Add decom script and install ps

holdenk committed:
Allow tail pid to fail

holdenk committed:
Backout the change to support non tarballs, this code path is more janky than I remember

holdenk committed:
Configure the memory of the decom, use specific batch interval so the first pod is always -1

holdenk committed:
Error message on invalid argument when running integratioon tests

holdenk committed:
Better error message when example jars are not present, fix case stement, and specify grace period since old fabric 8 defaults to 0

holdenk committed:
Give ourselves some time for attaching log

holdenk committed:
Try and fix the serialization issue when transfering blocks

holdenk committed:
Track exec pods better

holdenk committed:
Add terminating state

holdenk committed:
More logging

holdenk committed:
Integration tests pass

holdenk committed:
Fix up the mapstatus serialization

holdenk committed:
Log level cleanups

holdenk committed:
More minor cleanup

SparkQA commented:
**[Test build #132516 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/132516/testReport)** for PR 30692 at commit [`2921c59`](https://github.com/apache/spark/commit/2921c5990b98d28c262b14f07b2807cf696be72f).

SparkQA commented:
**[Test build #132516 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/132516/testReport)** for PR 30692 at commit [`2921c59`](https://github.com/apache/spark/commit/2921c5990b98d28c262b14f07b2807cf696be72f).
 * This patch **fails to generate documentation**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/37119/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/37119/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/37119/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/132516/

holdenk committed:
[SPARK-33724][K8S] Add decom script as a configuration param

Makes the location of the decommission script used in Kubernetes for graceful shutdown configurable.

Some environments don't use the Spark image builder and instead mount the decompressed Spark distro. In those envs configuring the location of the decommissioning script is required.

New configuration parameter.

Existing decommissioning integration test.

Closes #30694 from holdenk/SPARK-33724-allow-decommissioning-script-location-to-be-configured.

Authored-by: Holden Karau <hkarau@apple.com>
Signed-off-by: Dongjoon Hyun <dongjoon@apache.org>

SparkQA commented:
**[Test build #132586 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/132586/testReport)** for PR 30692 at commit [`941d7b0`](https://github.com/apache/spark/commit/941d7b0574211b34464ec2c3164c6417b7a002ca).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/37191/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/37191/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/37191/

SparkQA commented:
**[Test build #132586 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/132586/testReport)** for PR 30692 at commit [`941d7b0`](https://github.com/apache/spark/commit/941d7b0574211b34464ec2c3164c6417b7a002ca).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/132586/
